x-pack/plugins/apm/server/lib/apm_telemetry/collect_data_telemetry/tasks.ts

```diff
@@ -1150,4 +1151,180 @@ export const tasks: TelemetryTask[] = [
+              },
+              aggs: {
+                service_names: {
+                  terms: {
+                    field: SERVICE_NAME,
+                    size: 1000,
+                  },
+                  aggs: {
+                    top_metrics: {
+                      top_metrics: {
+                        sort: '_score',
+                        metrics: [
+                          {
+                            field: AGENT_NAME,
+                          },
+                          {
+                            field: AGENT_VERSION,
+                          },
+                          {
+                            field: SERVICE_LANGUAGE_NAME,
+                          },
+                          {
+                            field: SERVICE_LANGUAGE_VERSION,
+                          },
+                          {
+                            field: SERVICE_FRAMEWORK_NAME,
+                          },
+                          {
+                            field: SERVICE_FRAMEWORK_VERSION,
+                          },
+                          {
+                            field: SERVICE_RUNTIME_NAME,
+                          },
+                          {
+                            field: SERVICE_RUNTIME_VERSION,
+                          },
+                          {
+                            field: POD_NAME,
+                          },
+                          {
+                            field: CONTAINER_ID,
+                          },
+                        ],
+                      },
+                    },
+                    [CLOUD_REGION]: {
+                      terms: {
+                        field: CLOUD_REGION,
+                        size: 5,
+                      },
+                    },
+                    [CLOUD_PROVIDER]: {
+                      terms: {
+                        field: CLOUD_PROVIDER,
+                        size: 3,
+                      },
+                    },
+                    [CLOUD_AVAILABILITY_ZONE]: {
+                      terms: {
+                        field: CLOUD_AVAILABILITY_ZONE,
+                        size: 5,
+                      },
+                    },
+                    [FAAS_TRIGGER_TYPE]: {
+                      terms: {
+                        field: FAAS_TRIGGER_TYPE,
+                        size: 5,
+                      },
+                    },
+                  },
+                },
+              },
+            },
+          },
+        },
+      });
+      const data: APMPerService[] = [];
+      const envBuckets = response.aggregations?.environments.buckets ?? [];
+      envBuckets.forEach((envBucket) => {
+        const env = envBucket.key;
+        const serviceBuckets = envBucket.service_names?.buckets ?? [];
+        serviceBuckets.forEach((serviceBucket) => {
+          const name = serviceBucket.key;
+          const fullServiceName = env + '~' + name;
+          data.push({
```

Comment:
I think this can be immutable with `flatMap` and `map` instead of `forEach` and `push` (untested):

```suggestion
      const envBuckets = response.aggregations?.environments.buckets ?? [];
      const data = envBuckets.flatMap((envBucket) => {
        const { env } = envBucket;
        const serviceBuckets = envBucket.service_names?.buckets ?? [];
        return serviceBuckets.map((serviceBucket) => {
          const name = serviceBucket.key;
          const fullServiceName = `${env}~${name}`;
          return {
```

Reply:
Excellent addition, thank you!